tipuesearch_lite.js

```diff
@@ -152,34 +140,30 @@ window.onload = function execute(){
     }
 
     function selectPageText(pageText, searchTerms) {
-        let tempPageText = pageText.toLowerCase();
-        let posSearchTerm = -1;
-        for (const term of searchTerms) {
-            posSearchTerm = tempPageText.indexOf(term);
-            if (posSearchTerm != -1) {
-                break;
+        let tempText = pageText;
+        let bestText = "";
+        let score = 0;
+        // -1 to select always beginning of text by default
+        let bestScore = - 1;
+        while (tempText.length > 0) {
+            score = rateText(searchTerms, tempText.slice(0, set.contextLength));
+            if (score > bestScore) {
+                bestScore = score;
+                bestText = tempText.slice(0, set.contextLength);
             }
-        }
-        if (posSearchTerm > set.contextStart) {
-            let partialPageText = pageText.substr(posSearchTerm - set.contextBuffer);
-            partialPageText = pageText.substr(posSearchTerm - set.contextBuffer + partialPageText.indexOf(" "));
-            partialPageText = partialPageText.trim();
-            if (partialPageText.length > set.contextLength) {
-                pageText = "... " + partialPageText;
+            // check, otherwise endless loop
+            if (tempText.indexOf(" ") > - 1) {
+                tempText = tempText.slice(tempText.indexOf(" ") + 1);
+            } else {
+                tempText = "";
             }
         }
-        partialPageText = "";
-        let listOfPageText = pageText.split(" ");
-        if (listOfPageText.length < set.descriptiveWords) {
-            partialPageText = pageText;
-        } else {
-            partialPageText += listOfPageText.slice(0, set.descriptiveWords).join(" ");
-        }
-        partialPageText = partialPageText.trim();
-        if (partialPageText.charAt(partialPageText.length - 1) != ".") {
-            partialPageText += " ...";
+        if (pageText.length > set.contextLength && pageText.slice(0, set.contextLength) == bestText) {
```

Comment:
IMO, the logic is more easily read if nested rather than repeated. Also, the current approach has two problems:
- the content of `bestText` is altered after it has been scored. Content that is essential to that high score could be trimmed.
- it intends to remove partial words at the end, but it cannot tell the difference between partial words and complete words that don't have an extra space afterwards.

If the text is to be trimmed to the nearest complete word, then it should be done prior to scoring, assigned into `bestText`, and only manipulated to add `...` as appropriate.

For me, the following reads better:
```
if (pageText.length > set.contextLength) {
    if (pageText.slice(0, set.contextLength) == bestText) {
        bestText = bestText + " ...";
    } else {
        bestText = "... " + bestText + " ...";
    }
}
```